sharktank/sharktank/models/llm/testing.py

```diff
@@ -1,93 +1,164 @@
-            model.config.hp.context_length,
-        )
-        - 1,
-        dtype=torch.int64,
-        device=model.device,
-    )
+from copy import deepcopy
+from sharktank.types import InferenceTensorTransforms, Theta
+from sharktank.types.pipelining import pipeline_parallelize_llm_theta
+from sharktank.layers import LlamaModelConfig, ParallelismConfig
+from sharktank.utils.llm_utils import (
+    LlmPerplexityEval,
+    LlmInstance,
+    TorchInstance,
+    llama_config_page_sizes,
+    minimum_required_kv_cache_page_count_for_batch,
+)
+from sharktank.utils.testing import assert_tensor_close
+from typing import Callable
+
 
-    start_positions = [prefill_seq_lens]
-    seq_lens = prefill_seq_lens + 1
-    batch_seq_len = round_up_to_multiple_of(
-        int(torch.max(seq_lens)), model.config.block_seq_stride
+def make_random_token_sequences(
+    num_sequences: int,
+    min_tokens_per_sequence: int,
+    max_tokens_per_sequence: int,
+    vocabulary_size: int,
+) -> list[list[int]]:
+    assert min_tokens_per_sequence > 1
+    seq_lens = torch.randint(
+        size=[num_sequences],
+        low=min_tokens_per_sequence,
+        high=max_tokens_per_sequence + 1,
+        dtype=torch.int64,
     )
-    decode_token_ids = torch.randint(
+    max_seq_len = int(seq_lens.max())
+    token_ids = torch.randint(
         low=0,
-        high=model.config.hp.vocab_size,
-        size=[batch_size, 1],
-        dtype=torch.int32,
+        high=vocabulary_size,
+        size=[num_sequences, max_seq_len],
+        dtype=torch.int64,
     )
-    seq_block_ids = [
-        torch.arange(batch_size * batch_seq_len // model.config.block_seq_stride).view(
-            batch_size, -1
-        )
+    token_ids_list = token_ids.tolist()
+    token_ids_list = [
+        seq_token_ids[: int(seq_len)]
+        for seq_len, seq_token_ids in zip(seq_lens, token_ids_list)
     ]
-    cache_state = model.cache.allocate(page_count=seq_block_ids[0].numel() + batch_size)
-    cache_state = [torch.rand_like(cache_state[0])]
-    return OrderedDict(
-        [
-            ("tokens", decode_token_ids),
-            ("seq_lens", seq_lens),
-            ("start_positions", start_positions),
-            ("seq_block_ids", seq_block_ids),
-            ("cache_state", cache_state),
-        ]
-    )
+    return token_ids_list
 
 
-def make_random_prefill_args(
-    model: PagedLlmModelV1, batch_size: int
-) -> OrderedDict[str, Any]:
-    seq_lens = torch.randint(
-        size=[batch_size],
-        low=1,
-        high=min(
-            2 * model.config.block_seq_stride,
-            model.config.hp.context_length,
+class LlmPerplexityCompare:
+    """Compare the perplexity of one implementation against another.
+    Checks that the 2 models' results do not deviate for each other."""
+
+    def __init__(
+        self,
+        make_target_model: Callable[[], TorchInstance],
+        make_reference_model: Callable[[], TorchInstance],
```

Comment:
I don't understand these variable names. Why not `target_model` and `reference_model`?